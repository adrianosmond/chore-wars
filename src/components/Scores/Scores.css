.scores {
  margin-top: 2rem;
  margin-bottom: 2rem;
  position: relative;
}

.scores::before {
  content: "";
  position: absolute;
  height: 0.25rem;
  width: 100%;
  background-color: #ddd;
}

.scores__bar {
  position: absolute;
  right: 50%;
  height: 0.25rem;
  background-color: tomato;
  transition: width 1s;
}

.scores__bar--winning {
  background-image: linear-gradient(270deg, tomato 0, gold 100%);
}

.scores__bar--2 {
  left: 50%;
  right: auto;
}

.scores__bar--2.scores__bar--winning {
  background-image: linear-gradient(90deg, tomato 0, gold 100%);
}

.scores__difference {
  position: absolute;
  bottom: 0.5rem;
  padding: 0.25rem 0.375rem;
  font-size: 0.75rem;
  border-radius: 0.25rem;
  background-color: tomato;
  color: #fff;
  transform: translateX(var(--scores-difference-x-transform));
  transform-origin: 50% 100%;
}

.scores__difference::after {
  content: "";
  position: absolute;
  bottom: -3px;
  left: 50%;
  margin-left: -3px;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 3px 3px 0 3px;
  border-color: tomato transparent transparent transparent;
}

.scores__difference--1 {
  --scores-difference-x-transform: -50%;
  left: 1px;
}

.scores__difference--2 {
  --scores-difference-x-transform: 50%;
  right: 1px;
}

.scores__difference--winning {
  animation: jump 2s infinite;
  background-color: gold;
  color: #000;
}

.scores__difference--winning::after {
  border-color: gold transparent transparent transparent;
}

@keyframes jump {
  0% {
    transform: translateX(var(--scores-difference-x-transform));
  }
  15% {
    transform: translateX(var(--scores-difference-x-transform)) rotate(-10deg) translateY(-20%);
  }
  30% {
    transform: translateX(var(--scores-difference-x-transform));
  }
  45% {
    transform: translateX(var(--scores-difference-x-transform)) translateY(-20%);
  }
  60% {
    transform: translateX(var(--scores-difference-x-transform));
  }
  75% {
    transform: translateX(var(--scores-difference-x-transform)) rotate(10deg) translateY(-20%);
  }
  90% {
    transform: translateX(var(--scores-difference-x-transform));
  }
}

@media screen and (min-width: 37.5em) {
  .scores__difference {
    font-size: 1rem;
  }
}
