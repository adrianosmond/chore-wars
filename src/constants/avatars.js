export const avatarOptions = {
  skinColor: [
    'Pale',
    'Light',
    'Tanned',
    'Yellow',
    'Brown',
    'DarkBrown',
    'Black',
  ],
  topType: [
    'NoHair',
    'Eyepatch',
    'Hat',
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
    'LongHairBigHair',
    'LongHairBob',
    'LongHairBun',
    'LongHairCurly',
    'LongHairCurvy',
    'LongHairDreads',
    'LongHairFrida',
    'LongHairFro',
    'LongHairFroBand',
    'LongHairNotTooLong',
    'LongHairShavedSides',
    'LongHairMiaWallace',
    'LongHairStraight',
    'LongHairStraight2',
    'LongHairStraightStrand',
    'ShortHairDreads01',
    'ShortHairDreads02',
    'ShortHairFrizzle',
    'ShortHairShaggyMullet',
    'ShortHairShortCurly',
    'ShortHairShortFlat',
    'ShortHairShortRound',
    'ShortHairShortWaved',
    'ShortHairSides',
    'ShortHairTheCaesar',
    'ShortHairTheCaesarSidePart',
  ],
  hatColor: [
    'Black',
    'Blue01',
    'Blue02',
    'Blue03',
    'Gray01',
    'Gray02',
    'Heather',
    'PastelBlue',
    'PastelGreen',
    'PastelOrange',
    'PastelRed',
    'PastelYellow',
    'Pink',
    'Red',
    'White',
  ],
  hairColor: [
    'Auburn',
    'Black',
    'Blonde',
    'BlondeGolden',
    'Brown',
    'BrownDark',
    'PastelPink',
    'Platinum',
    'Red',
    'SilverGray',
  ],
  facialHairType: [
    'Blank',
    'BeardMedium',
    'BeardLight',
    'BeardMagestic',
    'MoustacheFancy',
    'MoustacheMagnum',
  ],
  facialHairColor: [
    'Auburn',
    'Black',
    'Blonde',
    'BlondeGolden',
    'Brown',
    'BrownDark',
    'Platinum',
    'Red',
  ],
  accessoriesType: [
    'Blank',
    'Kurt',
    'Prescription01',
    'Prescription02',
    'Round',
    'Sunglasses',
    'Wayfarers',
  ],
  clotheType: [
    'BlazerShirt',
    'BlazerSweater',
    'CollarSweater',
    'Hoodie',
    'Overall',
    'ShirtCrewNeck',
    'ShirtScoopNeck',
    'ShirtVNeck',
  ],
  clotheColor: [
    'Black',
    'Blue01',
    'Blue02',
    'Blue03',
    'Gray01',
    'Gray02',
    'Heather',
    'PastelBlue',
    'PastelGreen',
    'PastelOrange',
    'PastelRed',
    'PastelYellow',
    'Pink',
    'Red',
    'White',
  ],
};


export const editorOrder = [
  'skinColor',
  'topType',
  'hairColor',
  'hatColor',
  'facialHairType',
  'facialHairColor',
  'accessoriesType',
  'clotheType',
  'clotheColor',
];

export const labels = {
  skinColor: 'Skin Colour',
  topType: 'Hair or Headware',
  hairColor: 'Hair Colour',
  hatColor: 'Headware Colour',
  facialHairType: 'Facial Hair Type',
  facialHairColor: 'Facial Hair Colour',
  accessoriesType: 'Glasses',
  clotheType: 'Clothing Type',
  clotheColor: 'Clothing Colour',
};

export const isEditable = {
  skinColor: () => true,
  topType: () => true,
  hairColor: options => [
    'NoHair',
    'Eyepatch',
    'Hat',
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
    'LongHairFrida',
    'LongHairShavedSides',
  ].indexOf(options.topType) < 0,
  hatColor: options => [
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
  ].indexOf(options.topType) >= 0,
  facialHairType: options => options.topType !== 'Hijab',
  facialHairColor: options => isEditable.facialHairType(options) && options.facialHairType !== 'Blank',
  accessoriesType: () => true,
  clotheType: () => true,
  clotheColor: options => !options.clotheType.startsWith('Blazer'),
};

export const DefaultAvatar = {
  accessoriesType: 'Blank',
  avatarStyle: 'Transparent',
  clotheColor: 'Blue03',
  clotheType: 'BlazerShirt',
  eyeType: 'Default',
  eyebrowType: 'Default',
  facialHairColor: 'BrownDark',
  facialHairType: 'Blank',
  hairColor: 'BrownDark',
  mouthType: 'Twinkle',
  skinColor: 'Light',
  topType: 'LongHairStraight',
};
