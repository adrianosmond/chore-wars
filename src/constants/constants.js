const TIME_UNIT = 1000 * 60 * 60 * 24; // 1 day
const MIN_CHORE_FREQUENCY = 1;
const MAX_CHORE_FREQUENCY = 365;
const MIN_CHORE_POINTS = 1;
const MAX_CHORE_POINTS = 200;

const DATE_FORMAT = 'DD/MM/YYYY';

const MAX_POINT_DIFFERENCE = 500;

const MAX_NAME_LENGTH = 10;
const JOIN_CODE_LENGTH = 8;

const DefaultAvatar = {
  accessoriesType: 'Blank',
  avatarStyle: 'Transparent',
  clotheColor: 'Blue03',
  clotheType: 'BlazerShirt',
  eyeType: 'Default',
  eyebrowType: 'Default',
  facialHairColor: 'BrownDark',
  facialHairType: 'Blank',
  hairColor: 'BrownDark',
  mouthType: 'Twinkle',
  skinColor: 'Light',
  topType: 'LongHairStraight',
};

const ActionTypes = {
  addChore: 'ADD_CHORE',
  resetChoreDoneDate: 'RESET_CHORE_DONE_DATE',
  blockChore: 'BLOCK_CHORE',
  unblockChore: 'UNBLOCK_CHORE',
  removeChore: 'REMOVE_CHORE',
  updateChore: 'UPDATE_CHORE',
  setChores: 'SET_CHORES',
  makeChain: 'MAKE_CHAIN',
  breakChain: 'BREAK_CHAIN',
  addPoints: 'ADD_POINTS',
  claimPrize: 'CLAIM_PRIZE',
  paidDebt: 'PAID_DEBT',
  setPoints: 'SET_POINTS',
  updateUser: 'UPDATE_USER',
  setAuthUser: 'SET_AUTH_USER',
  setGame: 'SET_GAME',
  setPointsLoaded: 'SET_POINTS_LOADED',
  setChoresLoaded: 'SET_CHORES_LOADED',
  saveStatePostUndo: 'SAVE_STATE_POST_UNDO',
};

export {
  TIME_UNIT,
  MIN_CHORE_FREQUENCY,
  MAX_CHORE_FREQUENCY,
  MIN_CHORE_POINTS,
  MAX_CHORE_POINTS,
  MAX_NAME_LENGTH,
  JOIN_CODE_LENGTH,
  DATE_FORMAT,
  MAX_POINT_DIFFERENCE,
  ActionTypes,
  DefaultAvatar,
};
