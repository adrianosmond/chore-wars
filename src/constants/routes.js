export const LOGIN = '/'
export const CHORES = '/chores'
export const NEW_CHORE = '/chores/new'
