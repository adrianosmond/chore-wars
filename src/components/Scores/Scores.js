import React from 'react';
import { connect } from 'react-redux';
import Confetti from 'react-dom-confetti';

import { claimPrize } from 'actions/pointActions';

import { MAX_POINT_DIFFERENCE } from 'constants/constants';

import './Scores.css';

const Scores = (props) => {
  const { points } = props;
  const players = Object.keys(props.points);

  const minPoints = players.length === 2 ?
    Math.min(points[players[0]].points, points[players[1]].points) :
    points[players[0]].points;

  const scoresTied = players.length === 2 ?
    points[players[0]].points === points[players[1]].points :
    true;

  return (
    <div className="scores">
      {players.map((playerId, idx) => {
        const player = points[playerId];
        const playerBar = (player.points - minPoints);
        const winning = playerBar >= MAX_POINT_DIFFERENCE;
        return (
          <div className={`scores__bar scores__bar--${idx + 1}${winning ? ' scores__bar--winning' : ''}`}
            style={{ width: `${Math.min(50, (playerBar / MAX_POINT_DIFFERENCE) * 50)}%` }}
            key={playerId}>
            { playerBar || scoresTied ?
            <div className={`scores__difference scores__difference--${idx + 1}${winning ? ' scores__difference--winning' : ''}`}
              onClick={winning ? () => props.claimPrize(playerId, props.gameId) : null}>
              <Confetti active={!winning} />
              {playerBar}
              </div>
            : null }
          </div>
        );
      })}
    </div>
  );
};

const mapStateToProps = state => ({
  points: state.points.present,
  user: state.session.authUser.uid,
  gameId: state.session.game.gameId,
});

const mapDispatchToProps = dispatch => ({
  claimPrize: (player, game) => dispatch(claimPrize(player, game)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Scores);
