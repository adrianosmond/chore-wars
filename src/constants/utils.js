import { TIME_UNIT } from './constants'
const sortByCurrentPoints = (a, b) => b.currentPoints - a.currentPoints

const processChore = (chore) => {
  const timeSinceChore = (new Date().getTime() - chore.lastDone) / TIME_UNIT
  const timeRemaining = timeSinceChore - chore.frequency
  const percentage = Math.min(100 * timeSinceChore / chore.frequency, 100)
  const multiplier = 1 + (1 / chore.frequency)
  const due = chore.lastDone + (chore.frequency * TIME_UNIT)
  return {
    currentPoints: Math.round(percentage / 100 * chore.pointsPerTime * Math.pow(multiplier, timeRemaining)),
    percentage,
    due
  }
}

export { sortByCurrentPoints, processChore }
