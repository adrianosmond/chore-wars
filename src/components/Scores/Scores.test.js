import React from 'react';
import { shallow } from 'enzyme';
import { MAX_POINT_DIFFERENCE } from 'constants/constants';
import Scores from './Scores';

describe('Scores', () => {
  it('Renders an equal points graph', () => {
    expect(shallow(
      <Scores
        gameId="test-game"
        points={{
          player1: {
            name: 'Player 1',
            points: 100,
            isOwed: 0,
          },
          player2: {
            name: 'Player 2',
            points: 100,
            isOwed: 0,
          },
        }}
      />,
    )).toMatchSnapshot();
  });

  it('Renders a points graph with player 1 winning', () => {
    expect(shallow(
      <Scores
        gameId="test-game"
        points={{
          player1: {
            name: 'Player 1',
            points: 200,
            isOwed: 0,
          },
          player2: {
            name: 'Player 2',
            points: 100,
            isOwed: 0,
          },
        }}
      />,
    )).toMatchSnapshot();
  });

  it('Renders a points graph with player 2 winning', () => {
    expect(shallow(
      <Scores
        gameId="test-game"
        points={{
          player1: {
            name: 'Player 1',
            points: 100,
            isOwed: 0,
          },
          player2: {
            name: 'Player 2',
            points: 200,
            isOwed: 0,
          },
        }}
      />,
    )).toMatchSnapshot();
  });

  it('Renders a points graph with a difference of > MAX with p1 winning', () => {
    const p2Score = 100;
    expect(shallow(
      <Scores
        gameId="test-game"
        points={{
          player1: {
            name: 'Player 1',
            points: p2Score + MAX_POINT_DIFFERENCE + 100,
            isOwed: 0,
          },
          player2: {
            name: 'Player 2',
            points: p2Score,
            isOwed: 0,
          },
        }}
      />,
    )).toMatchSnapshot();
  });

  it('Renders a points graph with a difference of > MAX with p2 winning', () => {
    const p1Score = 100;
    expect(shallow(
      <Scores
        gameId="test-game"
        points={{
          player1: {
            name: 'Player 1',
            points: p1Score,
            isOwed: 0,
          },
          player2: {
            name: 'Player 2',
            points: p1Score + MAX_POINT_DIFFERENCE + 100,
            isOwed: 0,
          },
        }}
      />,
    )).toMatchSnapshot();
  });
});
