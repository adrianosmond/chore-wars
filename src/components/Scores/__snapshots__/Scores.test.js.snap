// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scores Renders a points graph with a difference of > MAX with p1 winning 1`] = `
<div
  className="scores"
>
  <div
    className="scores__bar scores__bar--1 scores__bar--winning"
    key="player1"
    style={
      Object {
        "width": "50%",
      }
    }
  >
    <div
      className="scores__difference scores__difference--1 scores__difference--winning"
      onClick={[Function]}
    >
      <Confetti
        active={false}
      />
      600
    </div>
  </div>
  <div
    className="scores__bar scores__bar--2"
    key="player2"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`Scores Renders a points graph with a difference of > MAX with p2 winning 1`] = `
<div
  className="scores"
>
  <div
    className="scores__bar scores__bar--1"
    key="player1"
    style={
      Object {
        "width": "0%",
      }
    }
  />
  <div
    className="scores__bar scores__bar--2 scores__bar--winning"
    key="player2"
    style={
      Object {
        "width": "50%",
      }
    }
  >
    <div
      className="scores__difference scores__difference--2 scores__difference--winning"
      onClick={[Function]}
    >
      <Confetti
        active={false}
      />
      600
    </div>
  </div>
</div>
`;

exports[`Scores Renders a points graph with player 1 winning 1`] = `
<div
  className="scores"
>
  <div
    className="scores__bar scores__bar--1"
    key="player1"
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <div
      className="scores__difference scores__difference--1"
      onClick={null}
    >
      <Confetti
        active={true}
      />
      100
    </div>
  </div>
  <div
    className="scores__bar scores__bar--2"
    key="player2"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`Scores Renders a points graph with player 2 winning 1`] = `
<div
  className="scores"
>
  <div
    className="scores__bar scores__bar--1"
    key="player1"
    style={
      Object {
        "width": "0%",
      }
    }
  />
  <div
    className="scores__bar scores__bar--2"
    key="player2"
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <div
      className="scores__difference scores__difference--2"
      onClick={null}
    >
      <Confetti
        active={true}
      />
      100
    </div>
  </div>
</div>
`;

exports[`Scores Renders an equal points graph 1`] = `
<div
  className="scores"
>
  <div
    className="scores__bar scores__bar--1"
    key="player1"
    style={
      Object {
        "width": "0%",
      }
    }
  />
  <div
    className="scores__bar scores__bar--2"
    key="player2"
    style={
      Object {
        "width": "0%",
      }
    }
  >
    <div
      className="scores__difference scores__difference--2"
      onClick={null}
    >
      <Confetti
        active={true}
      />
      0
    </div>
  </div>
</div>
`;
